puppet/lib/puppet/util/puppetdb/shared_terminus_state.rb

```diff
@@ -0,0 +1,32 @@
+
+require 'puppet/util/puppetdb/concurrent_hash_map'
+
+module Puppet::Util::Puppetdb
+  # A place to track the information which should be shared between all terminus
+  # instances running on a node - the index of the last URL we were able to use
+  # for querying. This should be as sticky as possible to prevent exported
+  # resource drift when master-puppetdb conections flicker.
+  #
+  # When running under puppet server, a real ConcurrentHashMap will be injected
+  # via the shared_terminus_state= class method which is shared between all
+  # local JRuby instances. When running under plain ruby, a local ruby
+  # implementation (called ConcurrentHashMap) is used.
+  class SharedTerminusState
+    @@shared_terminus_state = ConcurrentHashMap.new
+
+    @@MAP_KEY = :puppetdb_last_good_query_server_url_index
```

Comment:
Would be better to just use a constant here instead of a class variable; class variables are kind of evil